crates/turborepo-lib/src/daemon/connector.rs

```diff
@@ -0,0 +1,337 @@
+    ///
+    /// A new server will be spawned (and the old one killed) if
+    /// dont_kill is unset and one of these cases is hit:
+    /// 1. the versions do not match
+    /// 2. the server is not running
+    /// 3. the server is unresponsive
+    pub async fn connect(self) -> Result<DaemonClient<DaemonConnector>, DaemonConnectorError> {
+        let time = Instant::now();
+        for _ in 0..Self::CONNECT_RETRY_MAX {
+            let pid = self.get_or_start_daemon().await?;
+            debug!("got daemon with pid: {}", pid);
+
+            let conn = match self.get_connection(self.sock_file.clone()).await {
+                Err(DaemonConnectorError::Watcher(_) | DaemonConnectorError::Socket) => continue,
+                rest => rest?,
+            };
+
+            let mut client = DaemonClient {
+                client: conn,
+                connect_settings: (),
+            };
+
+            match client.handshake().await {
+                Ok(_) => {
+                    return {
+                        debug!("connected in {}ns", time.elapsed().as_micros());
+                        Ok(client.with_connect_settings(self))
+                    }
+                }
+                Err(DaemonError::VersionMismatch) if self.can_kill_server => {
+                    self.kill_live_server(client, pid).await?
+                }
+                Err(DaemonError::Unavailable) => self.kill_dead_server(pid).await?,
+                Err(_) => return Err(DaemonConnectorError::Handshake),
+            };
+        }
+
+        Err(DaemonConnectorError::Socket)
+    }
+
+    /// Gets the PID of the daemon process.
+    ///
+    /// If a daemon is not running, it starts one.
+    async fn get_or_start_daemon(&self) -> Result<sysinfo::Pid, DaemonConnectorError> {
+        debug!("looking for pid in lockfile: {:?}", self.pid_file);
+
+        let pidfile = self.pid_lock()?;
+
+        match pidfile.get_owner() {
+            Some(pid) => {
+                debug!("found pid: {}", pid);
+                Ok(sysinfo::Pid::from(pid as usize))
+            }
+            None if self.can_start_server => {
+                debug!("no pid found, starting daemon");
+                Self::start_daemon().await
+            }
+            None => Err(DaemonConnectorError::NotRunning),
+        }
+    }
+
+    /// Starts the daemon process, returning its PID.
+    async fn start_daemon() -> Result<sysinfo::Pid, DaemonConnectorError> {
+        let binary_path = std::env::current_exe().map_err(|_| DaemonConnectorError::Fork)?;
+
+        // this creates a new process group for the given command
+        // in a cross platform way, directing all output to /dev/null
+        let mut group = tokio::process::Command::new(binary_path)
+            .arg("daemon")
+            .stderr(Stdio::null())
+            .stdout(Stdio::null())
+            .group()
+            .kill_on_drop(false)
+            .spawn()
+            .map_err(|_| DaemonConnectorError::Fork)?;
```

Comment:
Here, and elsewhere, it looks like we're losing error details. Can we log at debug level what the error was so that we have somewhere to start when something isn't working?